File: gradle-json-verify/src/main/java/ninja/bryansills/JsonVerifyTask.java
package ninja.bryansills;

import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import okio.Okio;
import org.gradle.api.DefaultTask;
import org.gradle.api.file.ConfigurableFileTree;
import org.gradle.api.tasks.TaskAction;

import java.io.File;
import java.util.Map;

public class JsonVerifyTask extends DefaultTask {
    @TaskAction
    public void jsonVerify() {
        JsonVerifyExtension extension = (JsonVerifyExtension) getProject().getExtensions().findByName("jsonVerify");
        ConfigurableFileTree configFiles = getProject().fileTree(getProject().getProjectDir() + "/" + extension.getSrcDir());

        configFiles.forEach(file -> {
            if (".json".equals(getFileExtension(file))) {
                verifyJsonFile(file);
            }
        });
    }

    private void verifyJsonFile(File file) {
        Moshi moshi = new Moshi.Builder().build();
        JsonAdapter<Map> jsonAdapter = moshi.adapter(Map.class);

        try {
            Map jsonMap = jsonAdapter.fromJson(Okio.buffer(Okio.source(file)));
        } catch (Exception exception) {
            throw new JsonVerifyException(file.getName(), exception);
        }
    }

    private String getFileExtension(File file) {
        String name = file.getName();
        try {
            return name.substring(name.lastIndexOf("."));
        } catch (Exception e) {
            return "";
        }
    }
}

File: gradle-json-verify/src/main/java/ninja/bryansills/JsonVerifyTask.kt
package ninja.bryansills

import com.squareup.moshi.Moshi
import okio.Okio
import org.gradle.api.tasks.SourceTask
import org.gradle.api.tasks.TaskAction
import org.gradle.api.tasks.incremental.IncrementalTaskInputs
import java.io.File

open class JsonVerifyTask : SourceTask() {

    lateinit var output: File

    @TaskAction
    fun action(inputs: IncrementalTaskInputs) {
        inputs.outOfDate { inputFileDetails ->
            if (".json" == getFileExtension(inputFileDetails.file)) {
                verifyJsonFile(inputFileDetails.file)
            }
        }
    }

    private fun verifyJsonFile(file: File) {
        val moshi = Moshi.Builder().build()
        val jsonAdapter = moshi.adapter(Map::class.java)

        try {
            val jsonMap = jsonAdapter.fromJson(Okio.buffer(Okio.source(file)))
            println("${file.path} is good")
        } catch (exception: Exception) {
            throw JsonVerifyException("${exception.message} in file: ${file.path}", exception)
        }

    }

    private fun getFileExtension(file: File): String {
        val name = file.name

        return try {
            name.substring(name.lastIndexOf("."))
        } catch (e: Exception) {
            ""
        }

    }
}
